### PR TITLE
Implemented Preset change

### PR DESCRIPTION
**Pull request by thisguyStan**
- Implemented internal restart
- Implemented Service for presets
- Added Plugin for track changes

### CODE REVIEW

**Review by compujuckel — CHANGES_REQUESTED** [review]

File: CyclePresetPlugin/Preset/PresetManager.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+                Log.Information("Kicking all clients for preset change, server restart");
+                _entryCarManager.BroadcastPacket(new CSPKickBanMessageOverride { Message = RestartKickReason });
+                _entryCarManager.BroadcastPacket(new KickCar { SessionId = 255, Reason = KickReason.Kicked });
+            }
+        
+            var preset = new DirectoryInfo(CurrentPreset.UpcomingType!.PresetFolder).Name;
+        
+            // Restart the server
+            var sleep = (CurrentPreset.TransitionDuration - 1) * 1000;
+            Thread.Sleep(sleep);
```
[the remark this answers is not in the document]
Comment (thisguyStan) [reply CyclePresetPlugin/Preset/PresetManager.cs]: Changed method to async, changed Thread.Sleep to Task.Delay

File: CyclePresetPlugin/Preset/PresetConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        var deserializer = new DeserializerBuilder()
+            .IgnoreUnmatchedProperties()
+            .Build();
+
+        var yamlParser = new Parser(stream);
+        yamlParser.Consume<StreamStart>();
+        yamlParser.Accept<DocumentStart>(out _);
+
+        var cfg = deserializer.Deserialize<CyclePresetConfiguration>(yamlParser).Meta;
```
[the remark this answers is not in the document]
Comment (thisguyStan) [reply CyclePresetPlugin/Preset/PresetConfiguration.cs]: Simplified the deserialization 

https://github.com/thisguyStan/AssettoServer/blob/track-change/CyclePresetPlugin/Preset/PresetConfiguration.cs#L39-L41

File: CyclePresetPlugin/CyclePresetPlugin.cs [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+        if (_voteStarted)
+        {
+            context.Reply("Vote already ongoing.");
+            return;
+        }
+        
+        try
+        {
+            Log.Information("Starting preset vote");
+            VotingAsync(_cancellationToken, true).GetAwaiter().GetResult();
```
[the remark this answers is not in the document]
Comment (thisguyStan) [reply CyclePresetPlugin/CyclePresetPlugin.cs]: Unblocked the Task

File: CyclePresetPlugin/CyclePresetCommandModule.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    [Command("presetshow", "currentpreset", "currentrack"), RequireConnectedPlayer]
+    public void GetCurrentTrack()
+    {
+        _cyclePreset.GetTrack(Client!);
+    }
+
+    [Command("presetlist", "presetget", "presets"), RequireAdmin]
+    public void AdminTrackList()
+    {
+        _cyclePreset.ListAllPresets(Client!);
```
[the remark this answers is not in the document]
Comment (thisguyStan) [reply CyclePresetPlugin/CyclePresetCommandModule.cs]: Updated all Commands to use `BaseCommandContext` instead of `ACTcpClient`

**Comment by compujuckel** [comment]
Thanks!

**Pull request merged**